src/lib/salad/hll.h

```diff
@@ -0,0 +1,158 @@
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright 2010-2022, Tarantool AUTHORS, please see AUTHORS file.
+ */
+#pragma once
+
+#include <stdint.h>
+
+#if defined(__cplusplus)
+extern "C" {
+#endif /* defined(__cplusplus) */
+
+#include "lib/salad/hll_empirical.h"
+
+/**
+ * HyperLogLog supports sparse and dense represenations.
+ * The rerpesentation determines the data storage scheme.
+ */
+enum HLL_REPRESENTATION {
+	/**
+	 * Instead of registers the sparse representation works with pairs
+	 * consisting of the rank and the index of the added hashes. It requires
+	 * less memory for small cardinalities and provides the best accuracy.
+	 * Sparse representation swithces to dence representation if it starts
+	 * to require more amount of memory that is needed for the dense
+	 * representation.
+	 */
+	HLL_SPARSE,
+	/**
+	 * Dense representation allocates 2^precision counters. For small
+	 * cardinalities most counters are not used so this representation
+	 * should be used for esimating large cardinalities.
+	 */
+	HLL_DENSE
+};
+
+enum {
+	/** Precision of sparsely represented HyperLogLog. */
+	HLL_SPARSE_PRECISION = 26,
+};
+
+/**
+ * Estimator that is used for the HyperLogLog algorithm.
+ * The algorithm allows to estimate cardinality of a multiset using fixed amount
+ * of memory or even less. Memory requirements and estimation accuracy are
+ * determined by the algorithm precision parameter.
+ * For the dense representation the relative error is 1.04/sqrt(m) and the
+ * memory capasity is m*6 bits where m is number of counters wich
+ * equals to 2^precision.
+ * For the sparse representation the memory usage is proportional to the number
+ * of distinct elements that has added until it reaches the memory usage of the
+ * dense representation, and then it switches to the dense representation with
+ * fixed memory requretments. The sparse representation has the best accuracy.
+ */
+struct hll {
```

Comment:
The struct should be opaque (defined in the C file), since it isn't a part of the user API.

Reply:
Done.